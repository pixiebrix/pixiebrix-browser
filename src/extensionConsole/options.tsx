/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import "@/vendors/theme/app/app.scss";
import "@/vendors/overrides.scss";
import "@/utils/layout.scss";
import "./options.scss";

import "@/extensionContext";
import "@/development/darkMode";

import { render } from "react-dom";
import React from "react";
import App from "@/extensionConsole/App";
import initGoogle from "@/contrib/google/initGoogle";
import { initToaster } from "@/utils/notify";
import { initTelemetry } from "@/background/messenger/api";
import registerMessenger from "@/extensionConsole/messenger/registration";
import { initPerformanceMonitoring } from "@/telemetry/performance";

function init(): void {
  render(<App />, document.querySelector("#container"));
}

registerMessenger();
void initGoogle();
initToaster();
initTelemetry();
void initPerformanceMonitoring();
init();
