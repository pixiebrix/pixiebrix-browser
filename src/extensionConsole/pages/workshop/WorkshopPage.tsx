/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { useMemo, useState } from "react";
import Page from "@/layout/Page";
import { faHammer, faPlus, faTimes } from "@fortawesome/free-solid-svg-icons";
// eslint-disable-next-line no-restricted-imports -- TODO: Fix over time
import { Button, Form, InputGroup } from "react-bootstrap";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { isEmpty, orderBy, sortBy, uniq } from "lodash";
import Select from "react-select";
import workshopSlice, { type WorkshopState } from "@/store/workshopSlice";
import { connect, useDispatch, useSelector } from "react-redux";
import Fuse from "fuse.js";
import { push } from "connected-react-router";
import CustomBricksCard from "./CustomBricksCard";
import { type EnrichedBrick, type NavigateProps } from "./workshopTypes";
import { RequireScope } from "@/auth/RequireScope";
import { mapKindToKindUiValue } from "@/extensionConsole/pages/workshop/workshopUtils";
import { PACKAGE_REGEX } from "@/types/helpers";
import { useGetEditablePackagesQuery } from "@/services/api";
import { type EditablePackageMetadata } from "@/types/contract";

const { actions } = workshopSlice;

function selectRecent(state: { workshop: WorkshopState }) {
  return new Map((state.workshop.recent ?? []).map((x) => [x.id, x.timestamp]));
}

function selectFilters(state: { workshop: WorkshopState }) {
  return state.workshop.filters;
}

export function useEnrichBricks(
  bricks: EditablePackageMetadata[]
): EnrichedBrick[] {
  const recent = useSelector(selectRecent);

  return useMemo(() => {
    console.debug("Recent bricks", { recent });

    return orderBy(
      (bricks ?? []).map((brick) => {
        const match = PACKAGE_REGEX.exec(brick.name);
        return {
          ...brick,
          scope: match.groups.scope,
          collection: match.groups.collection,
          timestamp: recent.get(brick.id),
        };
      }),
      // Show recently accessed first
      [(x) => x.timestamp ?? -1, (x) => x.verbose_name],
      ["desc", "asc"]
    );
  }, [recent, bricks]);
}

export function useSearchOptions(bricks: EnrichedBrick[]) {
  const scopeOptions = useMemo(
    () =>
      sortBy(uniq((bricks ?? []).map((x) => x.scope))).map((value) => ({
        value,
        label: value ?? "[No Scope]",
      })),
    [bricks]
  );

  const collectionOptions = useMemo(
    () =>
      sortBy(uniq((bricks ?? []).map((x) => x.collection))).map((value) => ({
        value,
        label: value ?? "[No Collection]",
      })),
    [bricks]
  );

  const kindOptions = useMemo(
    () =>
      sortBy(uniq((bricks ?? []).map((x) => mapKindToKindUiValue(x.kind)))).map(
        (value) => ({
          value,
          label: value,
        })
      ),
    [bricks]
  );

  return {
    scopeOptions,
    collectionOptions,
    kindOptions,
  };
}

const CustomBricksSection: React.FunctionComponent<NavigateProps> = ({
  navigate,
}) => {
  const dispatch = useDispatch();
  const [query, setQuery] = useState("");
  const {
    data: remoteBricks,
    isLoading,
    error,
  } = useGetEditablePackagesQuery(undefined, {
    // Make sure user always see the latest bricks available (e.g., because they just saved a mod in the page editor)
    refetchOnMountOrArgChange: true,
  });
  const {
    scopes = [],
    collections = [],
    kinds = [],
  } = useSelector(selectFilters);
  const bricks = useEnrichBricks(remoteBricks);
  const { scopeOptions, kindOptions, collectionOptions } =
    useSearchOptions(bricks);

  const filtered = !isEmpty(scopes) || !isEmpty(collections) || !isEmpty(kinds);

  const fuse: Fuse<EnrichedBrick> = useMemo(
    () =>
      new Fuse(bricks, {
        keys: ["verbose_name", "name"],
      }),
    [bricks]
  );

  const sortedBricks = useMemo(() => {
    const results =
      query.trim() === "" ? bricks : fuse.search(query).map((x) => x.item);
    return results.filter(
      (x) =>
        (scopes.length === 0 || scopes.includes(x.scope)) &&
        (collections.length === 0 || collections.includes(x.collection)) &&
        (kinds.length === 0 || kinds.includes(mapKindToKindUiValue(x.kind)))
    );
  }, [fuse, query, scopes, collections, kinds, bricks]);

  return (
    <div className="max-950">
      <Form>
        <InputGroup className="mb-2 mr-sm-2">
          <InputGroup.Prepend>
            <InputGroup.Text>Search</InputGroup.Text>
          </InputGroup.Prepend>
          <Form.Control
            id="query"
            placeholder="Start typing to find results"
            value={query}
            onChange={({ target }) => {
              setQuery(target.value);
            }}
          />
        </InputGroup>
      </Form>

      <div className="d-flex">
        <div style={{ width: 200 }}>
          <Select
            isMulti
            placeholder="Filter @scope"
            options={scopeOptions}
            value={scopeOptions.filter((x) => scopes.includes(x.value))}
            onChange={(values) => {
              dispatch(actions.setScopes((values ?? []).map((x) => x.value)));
            }}
          />
        </div>
        <div style={{ width: 200 }} className="ml-3">
          <Select
            isMulti
            placeholder="Filter collection"
            options={collectionOptions}
            value={collectionOptions.filter((x) =>
              collections.includes(x.value)
            )}
            onChange={(values) => {
              dispatch(
                actions.setCollections((values ?? []).map((x) => x.value))
              );
            }}
          />
        </div>
        <div style={{ width: 200 }} className="ml-3">
          <Select
            isMulti
            placeholder="Filter type"
            options={kindOptions}
            value={kindOptions.filter((x) => kinds.includes(x.value))}
            onChange={(values) => {
              dispatch(actions.setKinds((values ?? []).map((x) => x.value)));
            }}
          />
        </div>
        <div className="ml-3">
          {filtered && (
            <Button
              variant="info"
              size="sm"
              style={{ height: "36px", marginTop: "1px" }}
              onClick={() => {
                dispatch(actions.clearFilters());
              }}
            >
              <FontAwesomeIcon icon={faTimes} /> Reset Filters
            </Button>
          )}
        </div>
      </div>
      <div className="mt-4">
        <CustomBricksCard
          navigate={navigate}
          bricks={sortedBricks}
          isFetching={isLoading}
          error={error}
        />
      </div>
    </div>
  );
};

const WorkshopPage: React.FunctionComponent<NavigateProps> = ({ navigate }) => (
  <RequireScope
    scopeSettingsTitle="Welcome to the PixieBrix Workshop!"
    scopeSettingsDescription="To use the Workshop, you must first set an account alias for your PixieBrix account"
  >
    <Page
      title="Workshop"
      icon={faHammer}
      description={
        <p>
          Text-based editor for advanced users to create and edit bricks. Mods
          created in the Page Editor can also be edited here.
        </p>
      }
      toolbar={
        <Button
          variant="info"
          onClick={() => {
            navigate("/workshop/create/");
          }}
        >
          <FontAwesomeIcon icon={faPlus} /> Create New Brick
        </Button>
      }
    >
      <CustomBricksSection navigate={navigate} />
    </Page>
  </RequireScope>
);

const mapDispatchToProps = { navigate: push };

export default connect(undefined, mapDispatchToProps)(WorkshopPage);
