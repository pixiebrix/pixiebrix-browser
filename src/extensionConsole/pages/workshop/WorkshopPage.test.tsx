/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import WorkshopPage, {
  useEnrichBricks,
  useSearchOptions,
} from "@/extensionConsole/pages/workshop/WorkshopPage";
import { render, renderHook } from "@/extensionConsole/testHelpers";
import { editablePackageMetadataFactory } from "@/testUtils/factories/registryFactories";
import React from "react";
import { screen } from "@testing-library/react";
import { authStateFactory } from "@/testUtils/factories/authFactories";
import { waitForEffect } from "@/testUtils/testHelpers";
import { authSlice } from "@/auth/authSlice";
import { appApiMock } from "@/testUtils/appApiMock";
import { MemoryRouter } from "react-router";

describe("useSearchOptions", () => {
  it("consolidates reader and block options", () => {
    const packages = [
      editablePackageMetadataFactory({ kind: "Reader" }),
      editablePackageMetadataFactory({ kind: "Block" }),
    ];

    const { result } = renderHook(() => {
      const bricks = useEnrichBricks(packages);
      return useSearchOptions(bricks);
    });
    expect(result.current).toEqual({
      collectionOptions: [{ label: "test", value: "test" }],
      kindOptions: [{ label: "Brick", value: "Brick" }],
      scopeOptions: [{ label: "[No Scope]", value: undefined }],
    });
  });
});

describe("WorkshopPage", () => {
  it("require account alias", () => {
    const { asFragment } = render(<WorkshopPage />);

    expect(
      screen.getByText(
        "To use the Workshop, you must first set an account alias for your PixieBrix account"
      )
    ).toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders empty workshop", async () => {
    appApiMock.reset();
    appApiMock.onGet("/api/bricks/").reply(200, []);

    const { asFragment } = render(
      <MemoryRouter>
        <WorkshopPage />
      </MemoryRouter>,
      {
        setupRedux(dispatch) {
          dispatch(
            authSlice.actions.setAuth(
              authStateFactory({
                scope: "workshopuser",
              })
            )
          );
        },
      }
    );

    await waitForEffect();

    expect(
      screen.queryByText(
        "To use the Workshop, you must first set an account alias for your PixieBrix account"
      )
    ).not.toBeInTheDocument();

    expect(screen.queryByTestId("loader")).not.toBeInTheDocument();
    expect(screen.getByText("No records found.")).toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });
});
