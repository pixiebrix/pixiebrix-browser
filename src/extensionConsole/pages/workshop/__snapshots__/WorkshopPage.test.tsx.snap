// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkshopPage renders empty workshop 1`] = `
<DocumentFragment>
  <div>
    <div
      class="d-flex"
    >
      <div
        class="flex-grow-1"
      >
        <div
          class="page-header"
        >
          <h3
            class="page-title"
          >
            <span
              class="page-title-icon bg-gradient-primary text-white mr-2"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-hammer fa-w-18 "
                data-icon="hammer"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 576 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M571.31 193.94l-22.63-22.63c-6.25-6.25-16.38-6.25-22.63 0l-11.31 11.31-28.9-28.9c5.63-21.31.36-44.9-16.35-61.61l-45.25-45.25c-62.48-62.48-163.79-62.48-226.28 0l90.51 45.25v18.75c0 16.97 6.74 33.25 18.75 45.25l49.14 49.14c16.71 16.71 40.3 21.98 61.61 16.35l28.9 28.9-11.31 11.31c-6.25 6.25-6.25 16.38 0 22.63l22.63 22.63c6.25 6.25 16.38 6.25 22.63 0l90.51-90.51c6.23-6.24 6.23-16.37-.02-22.62zm-286.72-15.2c-3.7-3.7-6.84-7.79-9.85-11.95L19.64 404.96c-25.57 23.88-26.26 64.19-1.53 88.93s65.05 24.05 88.93-1.53l238.13-255.07c-3.96-2.91-7.9-5.87-11.44-9.41l-49.14-49.14z"
                  fill="currentColor"
                />
              </svg>
            </span>
            Workshop
          </h3>
        </div>
      </div>
      <div>
        <button
          class="btn btn-info"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-plus fa-w-14 "
            data-icon="plus"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
              fill="currentColor"
            />
          </svg>
           Create New Brick
        </button>
      </div>
    </div>
    <div
      class="pb-4"
    >
      <p>
        Text-based editor for advanced users to create and edit bricks. Mods created in the Page Editor can also be edited here.
      </p>
    </div>
    <div
      class="max-950"
    >
      <form
        class=""
      >
        <div
          class="mb-2 mr-sm-2 input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="input-group-text"
            >
              Search
            </span>
          </div>
          <input
            class="form-control"
            id="query"
            placeholder="Start typing to find results"
            value=""
          />
        </div>
      </form>
      <div
        class="d-flex"
      >
        <div
          style="width: 200px;"
        >
          <div
            class=" css-b62m3t-container"
          >
            <span
              class="css-1f43avz-a11yText-A11yText"
              id="react-select-2-live-region"
            />
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              class="css-1f43avz-a11yText-A11yText"
            />
            <div
              class=" css-13cymwt-control"
            >
              <div
                class=" css-1fdsijx-ValueContainer"
              >
                <div
                  class=" css-1jqq78o-placeholder"
                  id="react-select-2-placeholder"
                >
                  Filter @scope
                </div>
                <div
                  class=" css-qbdosj-Input"
                  data-value=""
                >
                  <input
                    aria-autocomplete="list"
                    aria-describedby="react-select-2-placeholder"
                    aria-expanded="false"
                    aria-haspopup="true"
                    autocapitalize="none"
                    autocomplete="off"
                    autocorrect="off"
                    class=""
                    id="react-select-2-input"
                    role="combobox"
                    spellcheck="false"
                    style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                    tabindex="0"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class=" css-1hb7zxy-IndicatorsContainer"
              >
                <span
                  class=" css-1u9des2-indicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  class=" css-1xc3v61-indicatorContainer"
                >
                  <svg
                    aria-hidden="true"
                    class="css-tj5bde-Svg"
                    focusable="false"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ml-3"
          style="width: 200px;"
        >
          <div
            class=" css-b62m3t-container"
          >
            <span
              class="css-1f43avz-a11yText-A11yText"
              id="react-select-3-live-region"
            />
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              class="css-1f43avz-a11yText-A11yText"
            />
            <div
              class=" css-13cymwt-control"
            >
              <div
                class=" css-1fdsijx-ValueContainer"
              >
                <div
                  class=" css-1jqq78o-placeholder"
                  id="react-select-3-placeholder"
                >
                  Filter collection
                </div>
                <div
                  class=" css-qbdosj-Input"
                  data-value=""
                >
                  <input
                    aria-autocomplete="list"
                    aria-describedby="react-select-3-placeholder"
                    aria-expanded="false"
                    aria-haspopup="true"
                    autocapitalize="none"
                    autocomplete="off"
                    autocorrect="off"
                    class=""
                    id="react-select-3-input"
                    role="combobox"
                    spellcheck="false"
                    style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                    tabindex="0"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class=" css-1hb7zxy-IndicatorsContainer"
              >
                <span
                  class=" css-1u9des2-indicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  class=" css-1xc3v61-indicatorContainer"
                >
                  <svg
                    aria-hidden="true"
                    class="css-tj5bde-Svg"
                    focusable="false"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ml-3"
          style="width: 200px;"
        >
          <div
            class=" css-b62m3t-container"
          >
            <span
              class="css-1f43avz-a11yText-A11yText"
              id="react-select-4-live-region"
            />
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              class="css-1f43avz-a11yText-A11yText"
            />
            <div
              class=" css-13cymwt-control"
            >
              <div
                class=" css-1fdsijx-ValueContainer"
              >
                <div
                  class=" css-1jqq78o-placeholder"
                  id="react-select-4-placeholder"
                >
                  Filter type
                </div>
                <div
                  class=" css-qbdosj-Input"
                  data-value=""
                >
                  <input
                    aria-autocomplete="list"
                    aria-describedby="react-select-4-placeholder"
                    aria-expanded="false"
                    aria-haspopup="true"
                    autocapitalize="none"
                    autocomplete="off"
                    autocorrect="off"
                    class=""
                    id="react-select-4-input"
                    role="combobox"
                    spellcheck="false"
                    style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                    tabindex="0"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class=" css-1hb7zxy-IndicatorsContainer"
              >
                <span
                  class=" css-1u9des2-indicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  class=" css-1xc3v61-indicatorContainer"
                >
                  <svg
                    aria-hidden="true"
                    class="css-tj5bde-Svg"
                    focusable="false"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ml-3"
        />
      </div>
      <div
        class="mt-4"
      >
        <div
          class="card"
        >
          <div
            class="d-flex align-items-center card-header"
          >
            <div
              class="flex-grow-1"
            >
              Custom Bricks
            </div>
          </div>
          <div
            class="p-0 card-body"
          >
            <div
              class="table-responsive"
            >
              <table
                class="paginatedTable table"
                role="table"
                style="min-width: 0px;"
              >
                <thead
                  class="d-block w-100"
                >
                  <tr
                    role="row"
                    style="display: flex; flex: 1 0 auto; min-width: 0px;"
                  >
                    <th
                      class="th"
                      colspan="1"
                      role="columnheader"
                      style="position: relative; box-sizing: border-box; flex: 250 0 auto; min-width: 0px; width: 250px; cursor: pointer;"
                      title="Toggle SortBy"
                    >
                      <div
                        class="header-content d-flex"
                      >
                        Name
                        <span
                          class="text-muted ml-2"
                        >
                          <svg
                            aria-hidden="true"
                            class="svg-inline--fa fa-sort fa-w-10 sortIcon"
                            data-icon="sort"
                            data-prefix="fas"
                            focusable="false"
                            role="img"
                            viewBox="0 0 320 512"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41zm255-105L177 64c-9.4-9.4-24.6-9.4-33.9 0L24 183c-15.1 15.1-4.4 41 17 41h238c21.4 0 32.1-25.9 17-41z"
                              fill="currentColor"
                            />
                          </svg>
                        </span>
                      </div>
                      <div
                        class="resize"
                        draggable="false"
                        role="separator"
                        style="cursor: col-resize;"
                      />
                    </th>
                    <th
                      class="th"
                      colspan="1"
                      role="columnheader"
                      style="position: relative; box-sizing: border-box; flex: 150 0 auto; min-width: 0px; width: 150px; cursor: pointer;"
                      title="Toggle SortBy"
                    >
                      <div
                        class="header-content d-flex"
                      >
                        Collection
                        <span
                          class="text-muted ml-2"
                        >
                          <svg
                            aria-hidden="true"
                            class="svg-inline--fa fa-sort fa-w-10 sortIcon"
                            data-icon="sort"
                            data-prefix="fas"
                            focusable="false"
                            role="img"
                            viewBox="0 0 320 512"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41zm255-105L177 64c-9.4-9.4-24.6-9.4-33.9 0L24 183c-15.1 15.1-4.4 41 17 41h238c21.4 0 32.1-25.9 17-41z"
                              fill="currentColor"
                            />
                          </svg>
                        </span>
                      </div>
                      <div
                        class="resize"
                        draggable="false"
                        role="separator"
                        style="cursor: col-resize;"
                      />
                    </th>
                    <th
                      class="th"
                      colspan="1"
                      role="columnheader"
                      style="position: relative; box-sizing: border-box; flex: 150 0 auto; min-width: 0px; width: 150px; cursor: pointer;"
                      title="Toggle SortBy"
                    >
                      <div
                        class="header-content d-flex"
                      >
                        Type
                        <span
                          class="text-muted ml-2"
                        >
                          <svg
                            aria-hidden="true"
                            class="svg-inline--fa fa-sort fa-w-10 sortIcon"
                            data-icon="sort"
                            data-prefix="fas"
                            focusable="false"
                            role="img"
                            viewBox="0 0 320 512"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41zm255-105L177 64c-9.4-9.4-24.6-9.4-33.9 0L24 183c-15.1 15.1-4.4 41 17 41h238c21.4 0 32.1-25.9 17-41z"
                              fill="currentColor"
                            />
                          </svg>
                        </span>
                      </div>
                      <div
                        class="resize"
                        draggable="false"
                        role="separator"
                        style="cursor: col-resize;"
                      />
                    </th>
                    <th
                      class="th"
                      colspan="1"
                      role="columnheader"
                      style="position: relative; box-sizing: border-box; flex: 150 0 auto; min-width: 0px; width: 150px; cursor: pointer;"
                      title="Toggle SortBy"
                    >
                      <div
                        class="header-content d-flex"
                      >
                        Version
                        <span
                          class="text-muted ml-2"
                        >
                          <svg
                            aria-hidden="true"
                            class="svg-inline--fa fa-sort fa-w-10 sortIcon"
                            data-icon="sort"
                            data-prefix="fas"
                            focusable="false"
                            role="img"
                            viewBox="0 0 320 512"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41zm255-105L177 64c-9.4-9.4-24.6-9.4-33.9 0L24 183c-15.1 15.1-4.4 41 17 41h238c21.4 0 32.1-25.9 17-41z"
                              fill="currentColor"
                            />
                          </svg>
                        </span>
                      </div>
                    </th>
                  </tr>
                </thead>
                <tbody
                  class="d-block w-100"
                  role="rowgroup"
                >
                  <tr>
                    <td
                      class="text-muted"
                      colspan="5"
                    >
                      No records found.
                    </td>
                  </tr>
                </tbody>
                <tfoot
                  class="tfoot"
                >
                  <tr
                    class="d-block w-100"
                  >
                    <td
                      class="d-block w-100"
                    >
                      <div
                        class="d-flex align-items-center"
                      >
                        <div
                          class="text-muted"
                        >
                          Showing 0 to 0 of 0
                        </div>
                        <div
                          class="flex-grow-1"
                        >
                          <ul
                            class="m-0 float-right pagination"
                          >
                            <li
                              class="page-item disabled"
                            >
                              <span
                                class="page-link"
                                disabled=""
                              >
                                <span
                                  aria-hidden="true"
                                >
                                  «
                                </span>
                                <span
                                  class="sr-only"
                                >
                                  First
                                </span>
                              </span>
                            </li>
                            <li
                              class="page-item disabled"
                            >
                              <span
                                class="page-link"
                                disabled=""
                              >
                                <span
                                  aria-hidden="true"
                                >
                                  ‹
                                </span>
                                <span
                                  class="sr-only"
                                >
                                  Previous
                                </span>
                              </span>
                            </li>
                            <li
                              class="page-item disabled"
                            >
                              <span
                                class="page-link"
                                disabled=""
                              >
                                Page 1 of 1
                              </span>
                            </li>
                            <li
                              class="page-item disabled"
                            >
                              <span
                                class="page-link"
                                disabled=""
                              >
                                <span
                                  aria-hidden="true"
                                >
                                  ›
                                </span>
                                <span
                                  class="sr-only"
                                >
                                  Next
                                </span>
                              </span>
                            </li>
                            <li
                              class="page-item disabled"
                            >
                              <span
                                class="page-link"
                                disabled=""
                              >
                                <span
                                  aria-hidden="true"
                                >
                                  »
                                </span>
                                <span
                                  class="sr-only"
                                >
                                  Last
                                </span>
                              </span>
                            </li>
                          </ul>
                        </div>
                      </div>
                    </td>
                  </tr>
                </tfoot>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`WorkshopPage require account alias 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <div
      class="title"
    >
      Welcome to the PixieBrix Workshop!
    </div>
    <div
      class="font-weight-bold"
    >
      To use the Workshop, you must first set an account alias for your PixieBrix account
    </div>
    <div
      class="fade mt-2 alert alert-info show"
      role="alert"
    >
      <p>
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-info fa-w-6 "
          data-icon="info"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 192 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M20 424.229h20V279.771H20c-11.046 0-20-8.954-20-20V212c0-11.046 8.954-20 20-20h112c11.046 0 20 8.954 20 20v212.229h20c11.046 0 20 8.954 20 20V492c0 11.046-8.954 20-20 20H20c-11.046 0-20-8.954-20-20v-47.771c0-11.046 8.954-20 20-20zM96 0C56.235 0 24 32.235 24 72s32.235 72 72 72 72-32.235 72-72S135.764 0 96 0z"
            fill="currentColor"
          />
        </svg>
         Your account alias is a unique name used to prevent duplicate identifiers between the bricks you create and public/team bricks.
      </p>
    </div>
    <div
      class="fade mt-2 alert alert-info show"
      role="alert"
    >
      <p>
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-eye-slash fa-w-20 "
          data-icon="eye-slash"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 640 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M320 400c-75.85 0-137.25-58.71-142.9-133.11L72.2 185.82c-13.79 17.3-26.48 35.59-36.72 55.59a32.35 32.35 0 0 0 0 29.19C89.71 376.41 197.07 448 320 448c26.91 0 52.87-4 77.89-10.46L346 397.39a144.13 144.13 0 0 1-26 2.61zm313.82 58.1l-110.55-85.44a331.25 331.25 0 0 0 81.25-102.07 32.35 32.35 0 0 0 0-29.19C550.29 135.59 442.93 64 320 64a308.15 308.15 0 0 0-147.32 37.7L45.46 3.37A16 16 0 0 0 23 6.18L3.37 31.45A16 16 0 0 0 6.18 53.9l588.36 454.73a16 16 0 0 0 22.46-2.81l19.64-25.27a16 16 0 0 0-2.82-22.45zm-183.72-142l-39.3-30.38A94.75 94.75 0 0 0 416 256a94.76 94.76 0 0 0-121.31-92.21A47.65 47.65 0 0 1 304 192a46.64 46.64 0 0 1-1.54 10l-73.61-56.89A142.31 142.31 0 0 1 320 112a143.92 143.92 0 0 1 144 144c0 21.63-5.29 41.79-13.9 60.11z"
            fill="currentColor"
          />
        </svg>
         Your account alias will not be visible to anyone unless you choose to share a brick or mod.
      </p>
    </div>
    <div
      class="container"
    >
      <form
        class=""
        novalidate=""
      >
        <div
          class="formGroup form-group row"
        >
          <div
            class="mb-2 annotation-transition col-12"
          >
            <div
              class="annotationPlaceholder"
            />
          </div>
          <label
            class="label form-label col-form-label col-lg-auto"
            for="scope"
          >
            Account Alias
          </label>
          <div
            class="col-lg"
          >
            <input
              class="form-control"
              id="scope"
              name="scope"
              placeholder="@peter-parker"
              value=""
            />
            <small
              class="text-muted form-text"
            >
              <span>
                Your @alias for publishing bricks, e.g. @peter-parker
              </span>
            </small>
          </div>
        </div>
        <div
          class="submitDiv"
        >
          <button
            class="submit btn btn-primary"
            type="submit"
          >
            Set My Account Alias
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
